### PR TITLE
SONARJAVA-2674 Rule S3510: "HostnameVerifier.verify" should not always return true

### PR DESCRIPTION
**Pull request by christophe-zurn-sonarsource**

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/security/HostnameVerifierImplementationCheck.java [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+  private boolean isReturnTrueStatement(List<StatementTree> statementTreeList) {
+    if (statementTreeList.size() == 1 && statementTreeList.get(0).is(Tree.Kind.RETURN_STATEMENT)) {
+      ExpressionTree expression = ((ReturnStatementTree) statementTreeList.get(0)).expression();
+      return isTrueLiteral(expression);
+    }
+    return false;
+  }
+
+  private boolean isTrueLiteral(@Nullable ExpressionTree expressionTree) {
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/security/HostnameVerifierImplementationCheck.java]: Code Smell Code Smell: Make "isTrueLiteral" a "static" method. ([squid:S2325](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2325))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.java%3Ajava&pullRequest=1963&issues=AWJr5eRSRMG3VNtqdmej&open=AWJr5eRSRMG3VNtqdmej)

[ISSUE_KEY]: AWJr5eRSRMG3VNtqdmej

File: java-checks/src/main/java/org/sonar/java/checks/security/HostnameVerifierImplementationCheck.java [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    while (innerBlock.body().size() == 1 && innerBlock.body().get(0).is(Tree.Kind.BLOCK)) {
+      innerBlock = (BlockTree) innerBlock.body().get(0);
+    }
+
+    if (isReturnTrueStatement(innerBlock.body())) {
+      reportIssue(innerBlock.body().get(0), ISSUE_MESSAGE);
+    }
+  }
+
+  private boolean isHostnameVerifierSignature(LambdaExpressionTree lambdaExpressionTree) {
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/security/HostnameVerifierImplementationCheck.java]: Code Smell Code Smell: Make "isHostnameVerifierSignature" a "static" method. ([squid:S2325](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2325))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.java%3Ajava&pullRequest=1963&issues=AWJr5eRRRMG3VNtqdmei&open=AWJr5eRRRMG3VNtqdmei)

[ISSUE_KEY]: AWJr5eRRRMG3VNtqdmei

File: java-checks/src/main/java/org/sonar/java/checks/security/HostnameVerifierImplementationCheck.java [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+    } else if (tree.kind().equals(Tree.Kind.METHOD_INVOCATION)) {
+      ((MethodInvocationTree) tree).arguments().forEach(this::checkForLambdaImplementation);
+    } else {
+      ((NewClassTree) tree).arguments().forEach(this::checkForLambdaImplementation);
+    }
+  }
+
+  private void checkMethodDefinition(MethodTree tree) {
+    BlockTree blockTree = tree.block();
+    if (tree.symbol() != null && VERIFY_METHOD_MATCHER.matches(tree) && blockTree != null) {
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/security/HostnameVerifierImplementationCheck.java]: symbol cannot be null for a method tree.

File: java-checks/src/test/java/org/sonar/java/checks/security/HostnameVerifierImplementationCheckTest.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+package org.sonar.java.checks.security;
+
+import org.junit.Test;
+import org.sonar.java.checks.verifier.JavaCheckVerifier;
+
+public class HostnameVerifierImplementationCheckTest {
+
+  @Test
+  public void test() {
+    JavaCheckVerifier.verify("src/test/files/checks/security/HostnameVerifierImplementationCheck.java", new HostnameVerifierImplementationCheck());
```
Comment (benzonico) [inline java-checks/src/test/java/org/sonar/java/checks/security/HostnameVerifierImplementationCheckTest.java]: You may want to add a test without semantic (you should have no issue). There is a method for that on javaCheckVerifier.

File: java-checks/src/test/files/checks/security/HostnameVerifierImplementationCheck.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    return b.method(new HostnameVerifier() {
+      @Override
+      public boolean verify(String a, SSLSession b) {
+        return true; // Noncompliant {{Do not unconditionally return true in this method.}}
+      }
+    });
+  }
+
+  void method2() {
+    return b.method((a, b) -> true); // Noncompliant
```
Comment (benzonico) [inline java-checks/src/test/files/checks/security/HostnameVerifierImplementationCheck.java]: Your implementation won't catch a case where lambda is assigned to a variable. You should not check for lambda as argument but for every lambda of the proper type (hostnameVerifier) check if it is compliant or not.

File: java-checks/src/main/java/org/sonar/java/checks/security/HostnameVerifierImplementationCheck.java [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+  private static final TypeCriteria TYPE_CRITERIA_STRING = TypeCriteria.is("java.lang.String");
+  private static final TypeCriteria TYPE_CRITERIA_SSL_SESSION = TypeCriteria.is("javax.net.ssl.SSLSession");
+  private static final MethodMatcher VERIFY_METHOD_MATCHER = MethodMatcher.create().typeDefinition(TypeCriteria.subtypeOf("javax.net.ssl.HostnameVerifier"))
+    .name("verify")
+    .addParameter(TYPE_CRITERIA_STRING)
+    .addParameter(TYPE_CRITERIA_SSL_SESSION);
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return ImmutableList.of(Tree.Kind.METHOD, Tree.Kind.METHOD_INVOCATION, Tree.Kind.NEW_CLASS);
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/security/HostnameVerifierImplementationCheck.java]: If i'm correct you should end up with only Method and Lambda as subscribed kinds.

File: java-checks/src/main/java/org/sonar/java/checks/security/HostnameVerifierImplementationCheck.java [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+    }
+
+    if (isReturnTrueStatement(innerBlock.body())) {
+      reportIssue(innerBlock.body().get(0), ISSUE_MESSAGE);
+    }
+  }
+
+  private static boolean isHostnameVerifierSignature(LambdaExpressionTree lambdaExpressionTree) {
+    List<VariableTree> parameters = lambdaExpressionTree.parameters();
+    return lambdaExpressionTree.symbolType().isSubtypeOf("javax.net.ssl.HostnameVerifier")
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/security/HostnameVerifierImplementationCheck.java]: if you have a lambda as a subtype of HostnameVerifier I don't think you need all the elements of the condition. (in other words, there is no way if this first part is true that any of the following might be false, and it is therefore uncoverable by unit test and useless) :  so all the following && can be safely removed I think. (please prove me wrong if I am ;) )

File: java-checks/src/test/files/checks/security/HostnameVerifierImplementationCheck.java [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+  }
+
+  void method3() {
+    return b.method((a, b) -> (true)); // Noncompliant
```
Comment (benzonico) [inline java-checks/src/test/files/checks/security/HostnameVerifierImplementationCheck.java]: What about lambda : (a, b) -> (((true))) or return (((true))); 
I would add more than just one parenthesis.

File: java-checks/src/test/files/checks/security/HostnameVerifierImplementationCheck.java [hunk 9]
```diff
@@ -0,0 +1,72 @@
+import javax.net.ssl.HostnameVerifier;
```
Comment (benzonico) [inline java-checks/src/test/files/checks/security/HostnameVerifierImplementationCheck.java]: This one is an interface, I would add a testcase where this is implemented nicely and wrongly with a class definition. 

```
class A implements HostnameVerifier {
// wrong implemt
}
class B implements HostnameVerifier {
// correct implem
}

File: java-checks/src/test/files/checks/security/HostnameVerifierImplementationCheck.java (hunk 5 above)
Comment (benzonico) [inline java-checks/src/test/files/checks/security/HostnameVerifierImplementationCheck.java]: return is uncessary (and wrong) here.

File: java-checks/src/test/files/checks/security/HostnameVerifierImplementationCheck.java (hunk 8 above)
Comment (benzonico) [reply java-checks/src/test/files/checks/security/HostnameVerifierImplementationCheck.java]: I am not sure we need to put those cases in different methods (a bit more compact testcase could not hurt).

File: java-checks/src/test/files/checks/security/HostnameVerifierImplementationCheck.java [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+      }
+    });
+  }
+
+  void method9() {
+    new B((a, b) -> a.equalsIgnoreCase(b.getPeerHost()));
+  }
+}
+
+class B {
```
Comment (benzonico) [inline java-checks/src/test/files/checks/security/HostnameVerifierImplementationCheck.java]: Not sure what this testcase actually tests.

File: java-checks/src/main/java/org/sonar/java/checks/security/HostnameVerifierImplementationCheck.java [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+
+  private void checkForLambdaImplementation(ExpressionTree argument) {
+    if (argument.is(Tree.Kind.LAMBDA_EXPRESSION)) {
+      LambdaExpressionTree lambdaExpressionTree = ((LambdaExpressionTree) argument);
+      Tree lambdaBody = lambdaExpressionTree.body();
+      if (isHostnameVerifierSignature(lambdaExpressionTree)) {
+        if (lambdaBody.is(Tree.Kind.BLOCK)) {
+          checkBlock((BlockTree) lambdaBody);
+        } else if ((lambdaBody.is(Tree.Kind.PARENTHESIZED_EXPRESSION) || lambdaBody.is(Tree.Kind.BOOLEAN_LITERAL))
+          && isTrueLiteral((ExpressionTree) lambdaBody)) {
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/security/HostnameVerifierImplementationCheck.java]: Given the implem of isTrueLiteral skipping parenthesis there is no need for the first part of this &&. (in other words : you cannot fully cover this condition). the || should be removed.

**Pull request merged**